TAP-3.md

```diff
@@ -0,0 +1,217 @@
+- **State verification**: Verify the data integrity with a state proof. (this will be provided soon by Tycho Indexer directly.)
+- **Arbitrary depth**: Let users define the depth at which to probe the orderbook to define the mid-price. Denominated in the numeraire.
+- **Gas Price**: Include the cost of gas to execute the trade in the net price (and hence in the spread).
+- **Spread at depth**: Return also spread at chosen depth, in basis points. Gross and net gas (if gas feature is implemented otherwise just gross gas).
+- **N-hop parameter**: Let users define the maximum number of hops from token to numeraire token that the network search is allowed to take. Less hops will not cover some tokens, and might overlook some more liquid paths, but will be much faster. (absolute max should be no higher than 4 as the network will unreasonably explode here and might lead to bad UX if someone doesn't appreciate the implications of this setting).
+- **Cache for faster queries**: Cache all paths you calculated. And then only re-check the paths where at least on pool had an update since the block that this path was last updated.
+### Nice-to-have requirements
+- **Save price history**: Let users define tokens to track and calculate and save the price of these tokens to file at every block update, as long as the indexer is running. 
+- **Split across non-overlapping paths**: Instead of using the single path with the best price, use all parallel, non-overlapping, paths and split the trade efficiently between them, to receive a more precise quote.
+- **Calculate all prices**: Include a default setting where the app calculates and holds in memory all prices for all tokens it knows of. And updates them with every block continuously.
+- **Find best net price**: Instead of using a fixed depth, find the depth at which you can achieve the best net price (`net_price = (amount_out - gas_interms_of_token_out) / amount_in`) in both directions. Then take the mid-point between those prices.
+	- ![mid-price](./assets/mid-price.png)
+## Not included
+# Implementation
+You can derive the price for any token in terms of another in the following steps:
+1. Build a network from all available liquidity pools (nodes are tokens, pools are edges).
+2. Find all paths of max depth N (N-hop parameter) from numeraire to the target token.
+3. Swap M tokens (depth parameter) of numeraire to target token on all paths.
+4. (optional) Account for gas cost of the path (deduct from amount out).
+5. Pick the path with best (net gas) price.
+6. Swap the amount out given by the first swap, back into the other direction on the same path.
+7. Denote the of the prices from the two swaps, denominated in the numeraire token, as the tokens price.
+8. (optimal) Denote the mean difference of the swap prices at depth M to the mid-price (net gas), as the spread (denominated in basis points).
+## Design
+```ts
```

Comment:
Is there any requirement on what language this should be written? If not, I'd refrain from suggesting internal data types, or simply add a note saying that this is a suggestion.

Reply:
done